app/src/main/java/org/wikipedia/random/RandomViewModel.kt

```diff
@@ -0,0 +1,103 @@
+package org.wikipedia.random
+
+import android.os.Bundle
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.ViewModelProvider
+import androidx.lifecycle.viewModelScope
+import kotlinx.coroutines.*
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import org.wikipedia.Constants
+import org.wikipedia.WikipediaApp
+import org.wikipedia.analytics.RandomizerFunnel
+import org.wikipedia.database.AppDatabase
+import org.wikipedia.dataclient.WikiSite
+import org.wikipedia.page.PageTitle
+import org.wikipedia.util.log.L
+
+class RandomViewModel(bundle: Bundle) : ViewModel() {
+
+	private val funnel: RandomizerFunnel
+	private var saveShareJob: Job? = null
+
+	private val handler = CoroutineExceptionHandler { _, throwable ->
+		L.w(throwable)
+	}
+
+	private val _savaShareState = MutableStateFlow(false)
+	val savaShareState: StateFlow<Boolean> = _savaShareState
+
+	private val _saveToDefaultList = MutableStateFlow<SaveSharedState<PageTitle>>(Initial())
+	val saveToDefaultList: StateFlow<SaveSharedState<PageTitle>> = _saveToDefaultList
+
+	private val _saveToCustomList = MutableStateFlow<SaveSharedState<PageTitle>>(Initial())
+	val saveToCustomList: StateFlow<SaveSharedState<PageTitle>> = _saveToCustomList
+
+	private val _movePageToList = MutableStateFlow<SaveSharedState<MoveDataSource>>(Initial())
+	val movePageToList: StateFlow<SaveSharedState<MoveDataSource>> = _movePageToList
+
+	init {
+		val wikiSite = bundle.getParcelable<WikiSite>(RandomActivity.INTENT_EXTRA_WIKISITE)
+		val invokeSource =
+			bundle.getSerializable(Constants.INTENT_EXTRA_INVOKE_SOURCE) as Constants.InvokeSource
+		funnel = RandomizerFunnel(WikipediaApp.getInstance(), wikiSite, invokeSource)
+	}
+
+	fun saveToDefaultList(title: PageTitle) {
+		_saveToDefaultList.value = Result(title)
+	}
+
+	fun saveToCustomList(title: PageTitle) {
+		_saveToCustomList.value = Result(title)
+	}
+
+	fun movePageToList(sourceReadingListId: Long, title: PageTitle) {
+		_movePageToList.value = Result(MoveDataSource(sourceReadingListId, title))
+	}
```

Comment:
I'm not sure I see the benefit of the above three functions. You're passing in a PageTitle, which gets passed to a Flow, which gets passed right back to the Fragment. Why not keep these interactions within the Fragment as before?

Reply:
I'll fix it. I created these methods with the expectation that the logic could be complicated in the future